pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -64,18 +82,136 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
 	reqLogger.Info("Reconciling UserAccount")
 
 	// Fetch the UserAccount instance
-	instance := &toolchainv1alpha1.UserAccount{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	userAcc := &toolchainv1alpha1.UserAccount{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: config.GetOperatorNamespace(), Name: request.Name}, userAcc)
 	if err != nil {
 		if errors.IsNotFound(err) {
 			// Request object not found, could have been deleted after reconcile request.
 			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
 			// Return and don't requeue
 			return reconcile.Result{}, nil
 		}
-		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	var created bool
+	var user *userv1.User
+	if user, created, err = r.ensureUser(userAcc); err != nil || created == true {
+		return reconcile.Result{}, err
```

Comment:
Let's update the UserAccount status to `provisioning` if `ensureUser()` returns `created == true`.

Reply:
I guess, status=_provisioning_ will be done by host_cluster on MasterUserRecord which later sync to member_cluster with UserAccount.status=_provisioning_
Ref: [user provisioning seq flow](https://docs.google.com/document/d/19UzaXVqiOo9JqDYz3qUKguJwW-7zceA4UtTr358R_CA/edit#heading=h.bv9n6w3laifk)

Reply:
Host does not change status of MasterUserRecord directly.

T1. MasterUserRecord `state` is set to `approved`
T2. Host controller set its status to `provisioning` and creates a UserAccount in the member
T3. UserAccount controllers provisions the user and namespaces and updates the UserAccount status accordingly.
T4. When all done, the member controller updates the `resourceVersion` in the corresponding MasterUserAccount in host.
T5. Host controller checks the corresponding UserAccounts in member and updates the MasterUserRecord status.

So, each controller updates status of its primary resources only.

Reply:
> Let's update the UserAccount status to provisioning if ensureUser() returns created == true.

I think that this is too late - firstly, it should check if the User exists, if not, then it should set `status = provisioning` and in the next reconcile it should create all resources. Otherwise, it would create a `User` resource and then update the status which would trigger two reconciles

Reply:
I think that T3 step is more complex - we should break it down to:

T3-a: the UserAccount controller sets status = provisioning
T3-b: the member controller updates the resourceVersion in the corresponding MasterUserAccount in host.
T3-c: Host controller checks the corresponding UserAccounts in member and updates the MasterUserRecord status.
T3-d (in parallel with T3-c): UserAccount controllers provisions the user and namespaces and updates the UserAccount status accordingly.

Reply:
> Otherwise, it would create a User resource and then update the status which would trigger two reconciles

@MatousJobanek yes, I have come across this issue and used to get two reconcile callback.  Thanks for solution.

Overall, I see that coding to get `reconcile()` callback in proper manner is becoming complex and not obvious.  Each time you change some stuff and you need to see if you are getting proper `reconcile()` callback or not.  I really would like to check if other_operator has follow some different approach here.

Reply:
Hm.. I was assuming that status updates are ignored but they are not. Status updates do trigger reconciling.
But it looks like `GenerationChangedPredicate` from operator-sdk does the trick: https://github.com/operator-framework/operator-sdk/blob/121664e0093774c6d4fa4528fe7367d0848e6d74/pkg/predicate/predicate.go

I tested it and it looks like it filters out the events which caused by status updates only:
```
err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.MasterUserRecord{}}, &handler.EnqueueRequestForObject{}, predicate.GenerationChangedPredicate{})
...
```
In this case we can update the status without warring about triggering a reconcile.
Take a look at Knative serving controller for example: https://github.com/knative/serving-operator/blob/master/pkg/controller/knativeserving/knativeserving_controller.go#L142-L153

Reply:
@alexeykazakov this helped and working :+1:  Thank you :-)

Reply:
@MatousJobanek not sure if we want to do this as part of this PR for now https://github.com/codeready-toolchain/member-operator/pull/20#discussion_r297146111 @alexeykazakov  ??

Reply:
That’s fine to do the rest in separate PRs. Smaller PRs easier to review :)

Reply:
@MatousJobanek I guess there is already sub-task for https://github.com/codeready-toolchain/member-operator/pull/20#discussion_r297146111 .. task https://jira.coreos.com/browse/CRT-138

Reply:
Yes I agree, synchronization should not be part of this PR

Reply:
> ... But it looks like GenerationChangedPredicate from operator-sdk  ....

yeah, after reading this blog post https://blog.openshift.com/kubernetes-custom-resources-grow-up-in-v1-10/ I finally understand the purpose of the subresources and the usage.